packages/react-flicking/src/react-flicking/Flicking.tsx

```diff
@@ -1,188 +1,225 @@
-  }
+    this._diffResult = diffResult;
 
-  private checkCloneCount() {
-    const cloneCount = this.flicking!.getCloneCount();
+    const removedPanels = diffResult.removed.reduce((map, idx) => {
+      map[idx] = true;
+      return map;
+    }, {});
 
-    if (this.state.cloneCount !== cloneCount) {
-      this.setState({
-        cloneCount,
-      });
-    }
+    const panelsToRender = flicking.renderOnlyVisible
+      ? flicking.camera.visiblePanels
+      : flicking.renderer.panels;
+
+    this._panels = [
+      ...panelsToRender
+        .filter(panel => !removedPanels[panel.index])
+        // Sort panels by position
+        .sort((panel1, panel2) => (panel1.position + panel1.offset) - (panel2.position + panel2.offset))
+        .map(panel => diffResult.prevList[panel.index]),
+      ...diffResult.added.map(idx => diffResult.list[idx])
+    ];
+
+    return true;
   }
 
-  private renderPanels() {
-    const { renderOnlyVisible } = this.props;
-    const flicking = this.flicking;
-    const reactChildren = this.getChildren();
-    let panels: Array<React.ReactElement<any>>;
+  public componentDidUpdate() {
+    const flicking = this._nativeFlicking;
+    const renderer = flicking.renderer;
+    const diffResult = this._diffResult;
 
-    if (flicking && renderOnlyVisible) {
-      const diffResult = this.jsxDiffer.update(reactChildren.map(child => `${child.key}`));
-      const panelCnt = reactChildren.length;
+    if (!diffResult) return;
 
-      flicking.beforeSync(diffResult);
+    diffResult.removed.forEach(idx => {
+      renderer.remove(idx, 1);
+    });
 
-      const indexesToRender = flicking.getRenderingIndexes(diffResult);
-      panels = indexesToRender.map(index => {
-        if (index >= panelCnt) {
-          const relativeIndex = index % panelCnt;
-          const cloneIndex = Math.floor(index / panelCnt) - 1;
-          const origEl = reactChildren[relativeIndex];
+    diffResult.ordered.forEach(([prevIdx, newIdx]) => {
+      const prevPanel = renderer.panels[prevIdx];
+      const indexDiff = newIdx - prevIdx;
 
-          return <CloneComponent key={`clone${cloneIndex}${origEl.key}`}>{origEl}</CloneComponent>;
-        } else {
-          return reactChildren[index];
+      if (indexDiff > 0) {
+        prevPanel.increaseIndex(indexDiff);
+      } else {
+        prevPanel.decreaseIndex(-indexDiff);
+      }
+      // Update position
+      prevPanel.resize();
+    });
+
+    if (diffResult.added.length > 0) {
+      const children: HTMLElement[] = [].slice.call(this._cameraElement.children);
+      const addedElements = children.slice(-diffResult.added.length);
+      const cameraEl = this._cameraElement;
+
+      diffResult.added.forEach((panelIdx, elIdx) => {
+        const el = addedElements[elIdx];
+        const elNext = renderer.panels[panelIdx]
+          ? renderer.panels[panelIdx].element
+          : null;
+
+        if (el.nextElementSibling !== elNext) {
+          cameraEl.insertBefore(el, elNext);
```

Comment:
It is dangerous to add/remove/change DOM directly from the framework.

Reply:
That was the code before I update the panel's resize logic, I forgot to remove that!